### PR TITLE
Refactor specification to incorporate gittuf design

### PR DESCRIPTION
**Pull request by adityasaky**
I'm updating the spec document to incorporate the design. It includes the changes discussed in #13 and #15 and should guide the development of the implementation as we move away from the current prototype.

### CODE REVIEW

File: docs/specification.md [hunk 1]
```diff
@@ -1,49 +1,190 @@
...
+
+Last Modified: January 19, 2023
+
+Version: 0.1.0
+
+## Introduction
+
+This document describes gittuf, a security layer for Git repositories. gittuf
+applies several key properties part of the
+[The Update Framework (TUF)](https://theupdateframework.io/) such as delegations
```
Comment (JustinCappos) [inline docs/specification.md:11]:
```suggestion
[The Update Framework (TUF)](https://theupdateframework.io/) such as delegations,
secure key distribution, key revocation, trust rotation, read / write access control, 
```

File: docs/specification.md [hunk 2]
```diff
@@ -1,49 +1,190 @@
...
+and namespaces to Git repositories. This enables owners of the repositories to
+distribute (and revoke) contributor signing keys and define policies about which
+contributors can make changes to some namespaces within the repository. gittuf
+also implements the Reference State Log which was originally described in an
```
Comment (JustinCappos) [inline docs/specification.md:15]:
```suggestion
also protects against branch state attacks by extending the Reference State Log design
which was originally described in an
```

File: docs/specification.md [hunk 3]
```diff
@@ -1,49 +1,190 @@
...
+the repository.
+
+Policies for gittuf access are defined using a subset of TUF roles. The owners
+of the repository hold the keys used to sign the Root role that delegates trust
+to the other roles. The top level Targets role and any Targets roles it
+delegates to contain restrictions on protected namespaces. The specifics of the
+delegation structure vary from repository to repository as each will have its
+own constraints.
+
+A typical TUF delegation connects two TUF Targets role. In the delegations
```
Comment (JustinCappos) [inline docs/specification.md:97]:
```suggestion
A typical TUF delegation connects two TUF Targets roles. In the delegations
```

File: docs/specification.md [hunk 4]
```diff
@@ -1,49 +1,190 @@
...
+graph where each node is a Targets role, delegations are the edges connecting
+them. The delegations graph is traversed when verifying some target until the
+leaf node is encountered, after which the target's entry is compared against the
```
Comment (JustinCappos) [inline docs/specification.md:100]: This is a little confusing.  There really isn't a meaningful notion of leaf nodes as you may have cycles in delegations and may also have intermediate nodes that directly list targets in addition to delegating.

Comment (adityasaky) [reply docs/specification.md:102]: I'll take another pass. gittuf has a very opinionated approach to TUF delegations. Users are not expected to hand-craft metadata and we always expect the entries to be the commits themselves or the RSL entry so we shouldn't have the intermediate metadata recording hashes issue.

**Comment by adityasaky** [comment]
I've moved the SHA-256 bits to a standalone document. I see this going forward as:
1. primary specification that handles the TUF access control policies and corresponding workflows
2. other accompanying documents that describe extensions to this specification for hash algorithm agility, attestations, anything else we want to add

I believe the specification document is ready for reviews / discussions. We should tackle the SHA-256 bits separately.

File: docs/specification.md (hunk 4 above)
Comment (JustinCappos) [reply docs/specification.md:117]: What do you mean that entries are the commits themselves?  I thought they were either a path, branch, or tag?

Comment (adityasaky) [reply docs/specification.md]: I've updated this part of the document to try and make it clearer.

Comment (JustinCappos) [reply docs/specification.md]: I would tweak the new text to more state that TUF's namespacing is general and can be used for various git namespaces like tags, files, branches, etc.

Comment (adityasaky) [reply docs/specification.md]: Mentioned namespaces generally and said gittuf applies it to git but modifies the overall workflow to incorporate RSLs etc.

**Pull request merged**